core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -319,6 +318,15 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
 
       applications = appsToRetain
 
+      val modifiedTimesToRetain = new mutable.LinkedHashMap[String, Long]()
```

Comment:
What about simply deleting the entries that are too old? then the member can be `val` and doesn't need to be `@volatile`